File: src/main/kotlin/dev/slne/surf/essentialsold/listener/listeners/AdvancementListener.java
package dev.slne.surf.essentialsold.listener.listeners;

import dev.slne.surf.essentialsold.utils.EssentialsUtil;
import dev.slne.surf.essentialsold.utils.color.Colors;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.GameRule;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerAdvancementDoneEvent;

/**
 * A {@link Listener} for player advancement events, which sends a chat message to players when they complete an advancement.
 */
public class AdvancementListener implements Listener {
    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerAdvancementDone(PlayerAdvancementDoneEvent event) {
        val display = event.getAdvancement().getDisplay();
        val player = event.getPlayer();
        if (event.message() == null || display == null) return;

        if (Boolean.FALSE.equals(player.getWorld().getGameRuleValue(GameRule.ANNOUNCE_ADVANCEMENTS))) {
            return;
        }

        final String translationKey;

        switch (display.frame()) {
            case TASK -> translationKey = "chat.type.advancement.task";
            case CHALLENGE -> translationKey = "chat.type.advancement.challenge";
            default -> translationKey = "chat.type.advancement.goal";
        }

        event.message(EssentialsUtil.getPrefix()
                .append(Component.translatable(translationKey, EssentialsUtil.getDisplayName(player),
                                event.getAdvancement().displayName().colorIfAbsent(Colors.TERTIARY))
                        .color(Colors.GRAY)));
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/listener/AdvancementListener.kt
package dev.slne.surf.essentials.listener

import dev.slne.surf.essentials.util.translatable
import dev.slne.surf.surfapi.core.api.messages.Colors
import dev.slne.surf.surfapi.core.api.messages.adventure.buildText
import io.papermc.paper.advancement.AdvancementDisplay
import org.bukkit.GameRule
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerAdvancementDoneEvent

object AdvancementListener : Listener {
    @EventHandler
    fun onAdvancementDone(event: PlayerAdvancementDoneEvent) {
        val display = event.advancement.display ?: return
        val player = event.player

        if (event.message() == null) {
            return
        }

        if (player.world.getGameRuleValue(GameRule.ANNOUNCE_ADVANCEMENTS) == false) {
            return
        }

        val translationKey = when (display.frame()) {
            AdvancementDisplay.Frame.TASK -> "chat.type.advancement.task"
            AdvancementDisplay.Frame.CHALLENGE -> "chat.type.advancement.challenge"
            else -> "chat.type.advancement.goal"
        }

        event.message(buildText {
            appendPrefix()
            translatable(
                translationKey,
                player.displayName().colorIfAbsent(Colors.VARIABLE_VALUE),
                event.advancement.displayName().colorIfAbsent(Colors.YELLOW)
            ).colorIfAbsent(Colors.SPACER)
        })
    }
}